src/utils/usePromiseResolver.ts

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import {isPromise} from '../utils/guards';
+
+export const usePromiseResolver = <T>(mayBePromise: T | PromiseLike<T>): T | null => {
```

Comment:
nit: in the return value of this hook, there's no way to distinguish between a promise that resolved with `null` and a promise that has not yet resolved. It doesn't matter based on how this hook has been used in the `Elements` provider above, but I worry that this will be confusing in the future since it seems like this hook intends to be a general purpose utility.

How about changing the signature of this hook to be something like
```
<T>(maybePromise: T | PromiseLike<T>) => {settled: false} | {settled: true, value: T}
```

Reply:
I think we can go deeper and provide full promise state ;-)